server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,174 @@
+                        );
+                        if (includeDetails) {
+                            details.put("current_master", localMasterHistory.getCurrentMaster());
+                        }
+
+                    } else {
+                        logger.trace(
+                            String.format(
+                                Locale.ROOT,
+                                "This node thinks the master is unstable, but the master node %s thinks it is stable",
+                                master
+                            )
+                        );
+                        stableMasterStatus = HealthStatus.GREEN;
+                        summary = "The cluster has a stable master node";
+                    }
+                } catch (ExecutionException e) {
+                    logger.error("Exception trying to reach the master", e);
+                    stableMasterStatus = HealthStatus.YELLOW;
+                    summary = "The cluster has had a master node recently, but something went wrong while attempting to find out if the "
+                        + "master had been stable.";
+                    impacts.add(
+                        new HealthIndicatorImpact(
+                            3,
+                            "The cluster is at risk of not being able to create, delete, or rebalance indices",
+                            List.of(ImpactArea.INGEST)
+                        )
+                    );
+                } catch (InterruptedException e) {
+                    logger.info("Interrupted while trying to reach the master", e);
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
+                }
+            } else if (localMasterHistory.getDistinctMastersSeen().size() > 1 && localMasterHistory.getImmutableView().size() > 4) {
+                List<DiscoveryNode> mastersInLast30Minutes = localMasterHistory.getImmutableView();
+                logger.trace("Have seen " + (mastersInLast30Minutes.size() - 1) + " master changes in the last 30 seconds");
+                stableMasterStatus = HealthStatus.YELLOW;
+                summary = String.format(
+                    Locale.ROOT,
+                    "The master has changed %d times in the last 30 minutes",
+                    mastersInLast30Minutes.size() - 1
+                );
+                impacts.add(
+                    new HealthIndicatorImpact(
+                        3,
+                        "The cluster currently has a master node, but having multiple master node changes in a short time is an indicator "
+                            + "that the cluster is at risk of of not being able to create, delete, or rebalance indices",
+                        List.of(ImpactArea.INGEST)
+                    )
+                );
+                if (includeDetails) {
+                    details.put("current_master", localMasterHistory.getCurrentMaster());
```

Comment:
TODO: This causes XContent generation to blow up. Need to look into why, and to add a test.

Reply:
The reason is that XContentBuilder does not like ToXContentFragments (which is what DiscoveryNodes are) or nulls in lists (which we have here). I fixed it with a small wrapper class.